343/save users answers when submitting questionaire

richpjames opened the pull request:
## Context
The AP application form is fairly long and from research we have heard the current version can be completed over the course of a few days by probation practicioners. Our current implementation of sessions expire when the user is either logged out or closes their browser. It would be convinient for probation practicioners to be able partially fill in the form and then come back to it at a later time to complete it without having to keep the tab for the form open. This PR intends to allow that behaviour.

[Link to Trello ticket here](https://trello.com/c/w7eK5euE/343-save-the-users-answers-when-submitting-the-questionnare)

## Changes
Now when the `persistData` method of the `Form` class is called we write the session to a JSON file with the `sessionID` as it's file name.  
When the `initialize` method on `Form` is called (which is each time the user submits a response to a form page or loads a form page) we check whether the request has a body: 
- If it does have a body then the body is a new answer to a question so we persist that data. 
- If it doesn't have a body the call to `initialize` is a user loading a page so we check to see if there is a JSON file with the `sessionID` of the request and if there is we load their previous answers from JSON if they exist for that question. 

## Considerations
Saving the session to JSON is a short term solution to prove the concept. In beta stage the session should be saved to database.  
If, during development, the form changes (IE a question that previously required checkbox answers changes to a text field) this could cause errors when we load data from the session. In future we would like to be able to version the questions and then check that the saved responses are valid for that version of the questions but this work will be completed in a future PR.

Commit:
Add E2E tests for saving of responses

Here we run through the whole questionnaire, go back to the tasklist
and then go back through the questionnaire asserting that all the
responses we filled in last time have been selected. This is new
behaviour and proves the concept has worked.

Commit:
Add fs-extra package

this has some useful functions - outputFile & pathExists.

outputFile is preferable to the native (via ‘fs’ module) writeFile as
it creates folders in the path of the file if they don't exist.

pathExists is a tidier way of checking a path exists than is available in ‘fs’

Commit:
Read session from saved json file

‘form’s initialise method is either called with a request whose ‘body’
is the response to a question in the UI or a request with an empty
‘body’.

Here if the body doesn’t have any properties we attempt to read the
session from the saved json file we created earlier. To prevent
crashes we have to check the file exists before reading it. If the
‘body’ has no properties AND the saved session doesn’t exist we return
an empty object

Commit:
Write data to json when persisting

here we write the session to a json file with the sessionID as a file
name when the ‘persistData’ method is called. In the beta stage this
will probably be replaced with a write to the db but this proves the
concept for now.

I haven’t been able to work out why but sometimes ‘saveSession’ is
called with a function as the first argument. this call isn’t showing
up when logging or debugging but it does happen. For now to prevent it
 I have added a check that the ‘username’ is a string but these
erroneous calls should be investigated before/during beta.

Commit:
Move initialise tests into new 'describe' block

we will be adding to this block in a future commit

edavey commented on integration_tests/integration/referral-application.spec.ts:
```diff
@@ -78,6 +78,56 @@ context('SignIn', () => {
+
+    Page.verifyOnPage(ReferralApplicationTasklist)
+
+    // When I go back to previously answered sections
+    // Then my previous answers are visible
+    page.startSection('Check if the person is eligible')
+    page.checkInputsChecked('referralReason', 'likely')
+    cy.go('back')
```
Now that we're adding persistence outside of the "session", ought we to be a bit clearer in demonstrating how this work adds to the earlier session-based persistence? 

e.g. if we added an initial commit to this PR which showed the limits of the existing behaviour we're starting out with:

- When I fill in several steps
- And restart the questionnaire
- Then my answers are preserved
- When my session is destroyed
- Then my answers are no longer preserved

then the commits with the changes might follow on, with a final concluding commit like this one here which shows:

- When I fill in several steps
- And my session is destroyed
- And I restart the questionnaire
- Then my answers are preserved

I think it's the fact that we no longer rely on the session which is key change we want to highlight?

Would something of that sort work and be an improvement? Or have I got my wires crossed here?

richpjames replied:
I think that would work! I will have a go at it... Definitely a good way of making clear what the changes are. I also like the idea of having the E2E tests involved from the start of the PR

richpjames replied:
@pezholio I was hoping to log in/out but I currently can't do it locally

edavey reviewed (approved):
👍 nice work! I've made a couple of suggestions, which maybe we can discuss?

richpjames commented on server/forms/form.spec.ts:
```diff
@@ -130,13 +132,19 @@ describe('Form', () => {
             reason: 'likely',
           },
         },
+        sessionID: 'sessionID',
       })
 
       const form = await Form.initialize(request)
 
       form.persistData()
 
       expect(form.request.session.referralApplication).toEqual({ type: 'standard', reason: 'likely' })
+      expect(outputFile).toHaveBeenCalledWith(
+        `${__dirname}${savedSessionPath}sessionID.json`,
+        JSON.stringify(request.session.referralApplication),
+        'utf-8'
+      )
```
Yep, good idea

pezholio replied:
This could be clearer by having `MY-TOKEN` as a variable, then `join`-ing the variables in the expectation. Something like:

```typescript
const token = 'MY-TOKEN'

...

expect(outputFile).toHaveBeenCalledWith(
          `[__dirname, savedSessionPath, token '.json'].join(),
          JSON.stringify(request.session.referralApplication),
          'utf-8'
       )
```

pezholio commented on server/forms/form.ts:
```diff
@@ -58,4 +64,10 @@ export default class Form {
+  private static async writeJson(sessionID: string, data: any, sessionPath: string): Promise<void> {
+    if (typeof sessionID !== 'string') return
+    const stringifiedData = JSON.stringify(data)
+    await outputFile(`${__dirname}${sessionPath}${sessionID}.json`, stringifiedData, 'utf-8')
```
Again, `join`-ing might be neater here:

```typescript
const filepath = [__dirname, sessionPath, sessionID, '.json'].join()
```

Commit:
Add test case for ‘as is’ functionality

This demonstrates that when a user leaves the page their previously
entered answers aren’t saved

richpjames commented:
@pezholio ah yes, good spot 🤦‍♂️

pezholio commented on integration_tests/integration/referral-application.spec.ts:
```diff
@@ -79,6 +79,35 @@ context('SignIn', () => {
+    // Then my previous answers aren't visible
+    cy.get(`input[name="referralReason"][value=likely]`).should('not.be.checked')
+
+    cy.go('back')
```
I'm not quite sure I follow how these test case works out? How do these previously answered questions get cleared? And what does `cy.go(-3)` do?

I wonder if we could use `cy.session` with two named sessions (i.e. https://docs.cypress.io/api/commands/session#Conditionally-caching-a-session) to more easily demonstrate the fact that the answers disappear when the session is cleared.

richpjames replied:
I thought the answers being cleared was the expected behaviour. For me locally on `main` if I complete a section and then go back to it the answers aren't saved but the state on the Referral Tasklist shows Completed. 

cy.go(-3) is the same as clicking the browser's 'back' button 3 times - I opted to go back instead of completing the form because I thought it was clearer what was going on but I can see now that it would probably be more helpful to complete the form. https://docs.cypress.io/api/commands/go 

I like the idea of having two named sessions to demonstrate what happens when a session is wiped, I will have a go at that

Commit:
Update cypress to 9.6.0

to enable experimentalSessionAndOrigin which allows session functionality

pezholio reviewed (approved):
LGTM - thanks for persisting!
